docs/md/sink.md

```diff
@@ -0,0 +1,132 @@
+You can write workflow outputs to a Terra Workspace using the `Terra Workspace`
+sink. A typical `Terra Workspace` sink configuration in the workload request
+looks like:
+```json
+{
+  "name": "Terra Workspace",
+  "workspace": "{workspace-namespace}/{workspace-name}",
+  "entityType": "{name}",
+  "identifier": "{output-name}",
+  "fromOutputs": {
+    "attribute0": "output0",
+    "attribute1": ["output1", "output2"],
+    "attribute3": "$literal",
+    ...
+  }
+}
+``` 
+The table below summarises the purpose of each attribute in the above request.
+
+| Attribute     | Description                                                  |
+|---------------|--------------------------------------------------------------|
+| `name`        | Selects the `Terra Workspace` sink implementation.           |
+| `workspace`   | The Terra Workspace to write pipeline outputs to.            |
+| `entityType`  | The entity type in the `workspace` to write outputs to.      |
+| `identifier`  | Selects the output that will be used as the entity name.     |
+| `fromOutputs` | Mapping from outputs to attribute names in the `entityType`. |
+
+#### `workspace`
+The workspace is a `"{workspace-namespace}/{workspace-name}"` string as it
+appears in the URL path in the Terra UI. The workspace must exist prior to
+workload creation. You must ensure that `workflow-launcher@firecloud.org` is
+a workspace "Writer" in order to write entities to the workspace.
+
+#### `entityType`
+The `entityType` is the name of the entity type in the workspace that entities
+will be created as. The entity type must exist prior to workload creation and
+must be a table in the workspace.
+
+#### `identifier`
+The `identifier` is the name of a pipeline output that should be used as the 
+name of each newly created entity. 
+
+Example - Let's say the pipeline you're running has an output called
+"sample_name" that uniquely identifies the inputs and outputs to that pipeline.
+By setting `"identifier": "sample_name"` in the sink configuration, entities
+will be created using the "sample_name" as the entity name.
+
+Note that when two sets of pipeline outputs share the same "identifier" value,
+the first set of outputs will be overwritten by the second in the workspace.
+
+#### `fromOutputs`
+`fromOutputs` configures how to create new entities from pipeline outputs by 
+mapping the output names to attributes in the `entityType`. Note that all 
+attribute names must exist in the entityType before the workload  creation.
+
+`fromOutputs` allows a small about of flexibility in how to construct an entity
+and supports the following relations:
+
+- `"attribute": "output"`
+  Direct mapping from an output to an attribute
+
+- `"attribute": ["output0", "output2"]`
+  Make an attribute from a list of pipeline outputs.
+
+- `"attribute": "$value"`
+  Use the literal "value" for an attribute.
+
+## Developer Guide
+A sink is one satisfying the `Sink` protocol as below:
+```clojure
+(defprotocol Sink
```

Comment:
Sorry to say I'm confused by the use of a `defprotocol` here when we don't do this in the code :( with more distance from our work in the space, I'd see the mismatch between code and docs and assume that the documentation is out of date.

If a protocol-based approach helps in documentation, is it also something we can implement?  I'm not that familiar with them, doesn't look like they're widely used in `wfl`.

Reply:
I actually think we should use `defprotocol` because it forces us to define a data model for workload/source/sink/executor that's different from what a user sees and what's in the database.
The multi-methods are generic functions but the protocols define the operations you can perform on the set of objects of that type.

Reply:
Nice.  I agree this would be an improvement on implementation!

I'd feel good about approving this PR if we have a plan for imminent refactoring to `defprotocols` (maybe expand the scope of the "break apart COVID ns" tickets).  Otherwise I'm not in favor of releasing documentation that reflects an implementation change that we want to pursue but haven't.

Reply:
Sure - I'm keen to do it!

Reply:
I also like the protocols for illustrative purposes as they group together a set of operations and call that a thing whereas there's no way to really do this with multi-methods.